Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -236,111 +515,238 @@ void MultipleScatteringCorrection::exec() {
+    trackerLS1.reset(pos, vec);
+    trackerLS1.clearIntersectionResults();
+    const auto dist2 = getDistanceInsideObject(shape, trackerLS1);
+    //
+    LS1s[idx] = (dist1 - dist2);
+  }
 }
 
 /**
- * @brief parse and assign corresponding values from input properties
+ * @brief compute LS1s within given shape for sample and container case
  *
+ * @param distGraberContainer
+ * @param distGraberSample
+ * @param LS1sContainer
+ * @param LS1sSample
+ * @param shapeContainer
+ * @param shapeSample
  */
-void MultipleScatteringCorrection::parseInputs() {
-  // Get input workspace
-  m_inputWS = getProperty("InputWorkspace");
+void MultipleScatteringCorrection::calculateLS1s(const MultipleScatteringCorrectionDistGraber &distGraberContainer, //
+                                                 const MultipleScatteringCorrectionDistGraber &distGraberSample,    //
+                                                 std::vector<double> &LS1sContainer,                                //
+                                                 std::vector<double> &LS1sSample,                                   //
+                                                 const Geometry::IObject &shapeContainer,                           //
+                                                 const Geometry::IObject &shapeSample) const {
+  // Total elements = [container_elements] + [sample_elements]
+  // Schematic for scattering element i (*)
+  //   |                       \                                        /                       |
+  //   |      container         \               sample                 /  container             |
+  //   |                         \                                    /                         |
+  //   | ---LS1_container[i] ---  \  LS1_sample[i] * L2D_sample[i]   / ---L2D_container[i]  --- |
+  //   |                           \                                /                           |
+  const int64_t numVolumeElementsSample = distGraberSample.m_numVolumeElements;
+  const int64_t numVolumeElementsContainer = distGraberContainer.m_numVolumeElements;
+  const int64_t numVolumeElements = numVolumeElementsSample + numVolumeElementsContainer;
+  const auto sourcePos = m_inputWS->getInstrument()->getSource()->getPos();
+  Track trackerLS1(V3D{0, 0, 1}, V3D{0, 0, 1}); // reusable tracker for calculating LS1
+  for (int64_t idx = 0; idx < numVolumeElements; ++idx) {
+    const auto pos = idx < numVolumeElementsContainer
+                         ? distGraberContainer.m_elementPositions[idx]
+                         : distGraberSample.m_elementPositions[idx - numVolumeElementsContainer];
+    const auto vec = getDirection(sourcePos, pos);
+    //
+    trackerLS1.reset(sourcePos, vec);
+    trackerLS1.clearIntersectionResults();
+    const auto dist1_container = getDistanceInsideObject(shapeContainer, trackerLS1);
+    const auto dist1_sample = getDistanceInsideObject(shapeSample, trackerLS1);
```

Comment:
The second call (L598) will have the distances from L597 in addition to its own. You need to reset the tracker between these two calls otherwise `dist1_sample = <distance inside sample> + dist1_container`.

Reply:
fixing this section with suggested solution